### PR TITLE
Mark defaultTargetPlatform as constant for non-debug non-web builds.

### PR DESCRIPTION
**Pull request by sstrickl**
This PR adds the Dart VM `vm:platform-const-if` pragma introduced in https://github.com/dart-lang/sdk/commit/57a1168875 to the `defaultTargetPlatform` property, allowing it to be computed as if it was a constant field in non-debug AOT builds. In particular, this means that platform-specific code executed conditionally based on this property can be tree-shaken in release builds. Note that this PR changes `defaultTargetPlatform` to only allow overriding via `debugDefaultTargetPlatformOverride` in debug builds, and makes it so that compilation throws an error if code assigns to`debugDefaultTargetPlatformOverride` in other build modes.

Related issue: #14233

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/foundation/platform.dart [hunk 1]
```diff
@@ -94,5 +96,5 @@ enum TargetPlatform {
 /// certainly widgets to work assuming the presence of a system-wide back
 /// button, which will make those widgets unusable since iOS has no such button.
 ///
-/// In general, therefore, this property should not be used in release builds.
+/// Therefore this property can only override the platform in debug builds.
 TargetPlatform? debugDefaultTargetPlatformOverride;
```
[the remark this answers is not in the document]
Comment (sstrickl) [reply packages/flutter/lib/src/foundation/platform.dart]: Done.

**Comment by sstrickl** [comment]
Note that I haven't marked

> I added new tests to check the change I am making, or this PR is [test-exempt](https://github.com/flutter/flutter/wiki/Tree-hygiene#tests).

above, as I'm not sure what, if any, test would be useful for this.

**Review by mkustermann — APPROVED** [review]

File: packages/flutter/lib/src/foundation/platform.dart [hunk 2]
```diff
@@ -94,5 +97,17 @@ enum TargetPlatform {
...
-/// In general, therefore, this property should not be used in release builds.
-TargetPlatform? debugDefaultTargetPlatformOverride;
+/// Attempting to override this property in non-debug builds causes an error.
+TargetPlatform? get debugDefaultTargetPlatformOverride =>
+    _debugDefaultTargetPlatformOverride;
+
+set debugDefaultTargetPlatformOverride(TargetPlatform? value) {
+  if (!kDebugMode) {
+    throw FlutterError(
+      'Cannot modify debugDefaultTargetPlatformOverride in non-debug builds.\n'
```
Comment (Hixie) [inline packages/flutter/lib/src/foundation/platform.dart:107]: There's a stray `\n` at the end of the message.

**Comment by Hixie** [comment]
Is `vm:platform-const-if` documented anywhere?

**Comment by Hixie** [comment]
test-exempt: optimization

**Comment by mkustermann** [comment]
> Is vm:platform-const-if documented anywhere?

All VM related pragmas are documented at [runtime/docs/pragmas.md](https://github.com/dart-lang/sdk/blob/main/runtime/docs/pragmas.md)

**Comment by Hixie** [comment]
@matanlurey points out (correctly) that it'd be nice to test that in production mode we do throw for setting the debug flag here. do we have a harness for testing things in production mode? if we don't i think this probably doesn't justify the cost of adding one, but if we do, it'd be nice to add this to it.

File: packages/flutter/lib/src/foundation/platform.dart (hunk 2 above)
Comment (sstrickl) [reply packages/flutter/lib/src/foundation/platform.dart]: Thanks, done!

**Comment by sstrickl** [comment]
> @matanlurey points out (correctly) that it'd be nice to test that in production mode we do throw for setting the debug flag here. do we have a harness for testing things in production mode? if we don't i think this probably doesn't justify the cost of adding one, but if we do, it'd be nice to add this to it.

I'll take a look around and see if I find one, and if I do then add it before merging.

**Comment by sstrickl** [comment]
I didn't find a harness for testing non-debug-only code when I looked around, and from Discord it seems like no one else knows of one either to point me at, so I'm going to ship this as is.

**Comment by mkustermann** [comment]
> Hi @sstrickl, just an fyi that this PR is causing post submit [errors](https://flutter-dashboard.appspot.com/)
>
> I'll be investigating some more to see if your PR is causing any issues related to the observed error

@eliasyishak To follow up here, were there any issues we need to be aware of?

**Comment by sstrickl** [comment]
Elias reached out to me over chat and it turned out to be a flake after further investigation, hence why no revert happened.

**Pull request merged**